bofire/surrogates/mixed_tanimoto_gp.py

```diff
@@ -0,0 +1,369 @@
+import base64
+import io
+import warnings
+from functools import partial
+from typing import Callable, Dict, List, Optional
+
+import pandas as pd
+import torch
+from botorch.fit import fit_gpytorch_mll
+from botorch.models.gp_regression import SingleTaskGP
+from botorch.models.kernels.categorical import CategoricalKernel
+from botorch.models.transforms.input import (
+    ChainedInputTransform,
+    InputTransform,
+    OneHotToNumeric,
+)
+from botorch.models.transforms.outcome import OutcomeTransform, Standardize
+from botorch.utils.transforms import normalize_indices
+from gpytorch.constraints import GreaterThan
+from gpytorch.kernels.kernel import Kernel
+from gpytorch.kernels.matern_kernel import MaternKernel
+from gpytorch.kernels.scale_kernel import ScaleKernel
+from gpytorch.likelihoods.gaussian_likelihood import GaussianLikelihood
+from gpytorch.likelihoods.likelihood import Likelihood
+from gpytorch.mlls import ExactMarginalLogLikelihood
+from gpytorch.priors import GammaPrior
+from torch import Tensor
+
+import bofire.kernels.api as kernels
+from bofire.data_models.enum import (
+    CategoricalEncodingEnum,
+    OutputFilteringEnum,
+)
+from bofire.data_models.molfeatures.api import (
+    Fingerprints,
+    FingerprintsFragments,
+    Fragments,
+    MordredDescriptors,
+)
+
+# MolecularEncodingEnum,
+from bofire.data_models.surrogates.api import MixedTanimotoGPSurrogate as DataModel
+
+# from bofire.data_models.kernels.categorical import HammondDistanceKernel
+from bofire.surrogates.botorch import BotorchSurrogate
+from bofire.surrogates.single_task_gp import get_scaler
+from bofire.surrogates.trainable import TrainableSurrogate
+from bofire.utils.torch_tools import tkwargs
+
+
+class MixedTanimotoGP(SingleTaskGP):
+    def __init__(
+        self,
+        train_X: Tensor,
+        train_Y: Tensor,
+        mol_dims: List[int],
+        mol_kernel_factory: Callable[[torch.Size, int, List[int]], Kernel],
+        cat_dims: Optional[List[int]] = None,
+        # cat_kernel_factory: Optional[
+        #    Callable[[torch.Size, int, List[int]], Kernel]
+        # ] = None, --> BoTorch forced to use CategoricalKernel
+        cont_kernel_factory: Optional[
+            Callable[[torch.Size, int, List[int]], Kernel]
+        ] = None,
+        likelihood: Optional[Likelihood] = None,
+        outcome_transform: Optional[OutcomeTransform] = None,
+        input_transform: Optional[InputTransform] = None,
+    ) -> None:
+        if len(mol_dims) == 0:
+            raise ValueError("Must specify molecular dimensions for MixedTanimotoGP")
+
+        if cat_dims is None:
```

Comment:
```suggestion
        cat_dims = cat_dims or []
```

Reply:
I have made the changes based on this suggestion.